keps/sig-scheduling/4187-a-generic-scheduler-cache-extension/README.md

```diff
@@ -0,0 +1,777 @@
+
+Moving the cache creation/management logic, list/watch mechanism, event handling flow as well common data structures to a centralized cache extension will simplify the scheduler plugin developer’s life and avoid data inconsistency issue.
+
+### Goals
+
+- Make the scheduler easier to be extended for complicated resource handling
+- Simplify the scheduler plugin development
+
+### Non-Goals
+
+- Replace the device plugin mechanism
+- Convert existing scheduler plugins to use the cache extension
+
+## Proposal
+
+The proposal is to extend the scheduler’s cache for scheduler plugins with customized extended resources. It introduces a generic extension to the scheduler cache to hold extended resources, which includes resource telemetry data gathered from the low-level metrics collectors. It also introduces new interfaces and methods to manage the life cycle of the extended resources, such as the resource creation, update, and deletion. It also defines user-friendly APIs for the scheduler plugin developers to describe resources and customize their counting methods.
+
+### User Stories
+
+#### Disk IO Aware Scheduling
+
+There are emerging requirements to guarantee critical workloads’ disk IO performance. Current K8S scheduler only provides CPU/RAM and ephemeral storage aware scheduling, but it does not provide disk io resource scheduling. When we use existing scheduler framework to develop the disk IO aware scheduler plugin, we need to  create a separate cache in the scheduler plugin to save the disk IO related information such as available capacity and re-implement the pod/node events handling logic. Due to the independent event handling processes, different scheduling plugins (e.g. in-tree scheduling plugin and the customized out-of-tree scheduling plugin) may make the scheduling decision based on resource information snapshotted at different time points, which may result in unnecessary scheduling failure. For instance, right after pod1 is created on node1, pod1 is deleted. Suppose that the in-tree NodeResourcesFit plugin gets the deletion notification at time t1 and the out-of-tree disk io plugin gets the deletion notification at time t2. It could happen that the two scheduler plugins make their scheduling decision at a time point between t1 and t2. In this case, the resource availability information based on which the two scheduler plugins make their decisions would diverge, which may result in scheduling failure or unnecessary pod preemption. With the cache extension, the first-class resource used by NodeResourcesFit plugin and the io resource used by disk io plugin are updated through one event notification and there will be no divergence of resource information. This reduces the probability of pod rescheduling or preemption.
```

Comment:
> Is the example here related to disk IO aware scheduling? I can imagine pod affinity might be impacted by the pod deletion (pod deletion can lead to a scheduling failure), but how the event impacts the disk IO?

@chendave  Thanks for providing comments. Disk IO is a typical vendor-specific resource. Thus disk IO aware scheduling involves an out-of-tree scheduler plugin which caches its resource info for making scheduling decision. 
Imagine a scenario where the NodeResourcesFit plugin(in-tree) receives a Pod deletion notification at a moment we’ll call t1. Simultaneously, the disk IO plugin(out-of-tree) is notified of the Pod deletion at a different moment, t2. There’s a possibility that both scheduler plugins could make their scheduling decisions at a time point between t1 and t2. The in-tree plugin received the pod deletion event while the out-of-tree plugin did not. It would lead to scheduling failure.

Reply:
> btw, is the `Disk IO Aware Scheduling` has already be open sourced in the scheduler plugin or any plan to do so?

Here is the [on-going PR](https://github.com/kubernetes-sigs/scheduler-plugins/pull/628) in scheduler plugins.